### ISSUE TITLE
Unable to build wheel from unittest2

### ISSUE THREAD

**Issue by ghost** [pypa/setuptools#409]
Originally reported by: **msabramo (Bitbucket: [msabramo](http://bitbucket.org/msabramo), GitHub: [msabramo](http://github.com/msabramo))**

---

Because unittest2 has a version that is an object (an instance of `late_version`); not a `str` as setuptools is expecting:

``` pytb
$ mkvirtualenv unittest2-wheel-test
New python executable in unittest2-wheel-test/bin/python
Please make sure you remove any previous custom paths from your /Users/marca/.pydistutils.cfg file.
Installing setuptools, pip, wheel...done.
Python 2.7.9
setuptools 17.0
pip 7.0.3 from /Users/marca/python/virtualenvs/unittest2-wheel-test/lib/python2.7/site-packages (python 2.7)

$ pip wheel --no-use-wheel unittest2
...
Building wheels for collected packages: unittest2, argparse, six, traceback2, linecache2
  Running setup.py bdist_wheel for unittest2
  Complete output from command /Users/marca/python/virtualenvs/unittest2-wheel-test/bin/python -c "import setuptools;__file__='/private/var/folders/gw/w0clrs515zx9x_55zgtpv4mm0000gp/T/pip-build-wyUgjq/unittest2/setup.py';exec(compile(open(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" bdist_wheel -d /var/folders/gw/w0clrs515zx9x_55zgtpv4mm0000gp/T/tmpNXMSLVpip-wheel-:
  /Users/marca/python/virtualenvs/unittest2-wheel-test/lib/python2.7/site-packages/setuptools/dist.py:294: UserWarning: The version specified (<__main__.late_version instance at 0x102d81368>) is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
    "details." % self.metadata.version
  running bdist_wheel
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/private/var/folders/gw/w0clrs515zx9x_55zgtpv4mm0000gp/T/pip-build-wyUgjq/unittest2/setup.py", line 87, in <module>
      setup(**params)
    File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/core.py", line 151, in setup
      dist.run_commands()
    File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 953, in run_commands
      self.run_command(cmd)
    File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/dist.py", line 971, in run_command
      cmd_obj.ensure_finalized()
    File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/cmd.py", line 109, in ensure_finalized
      self.finalize_options()
    File "/Users/marca/python/virtualenvs/unittest2-wheel-test/lib/python2.7/site-packages/wheel/bdist_wheel.py", line 106, in finalize_options
      self.data_dir = self.wheel_dist_name + '.data'
    File "/Users/marca/python/virtualenvs/unittest2-wheel-test/lib/python2.7/site-packages/wheel/bdist_wheel.py", line 127, in wheel_dist_name
      safer_version(self.distribution.get_version())))
    File "/Users/marca/python/virtualenvs/unittest2-wheel-test/lib/python2.7/site-packages/wheel/bdist_wheel.py", line 47, in safer_version
      return safe_version(version).replace('-', '_')
    File "/Users/marca/python/virtualenvs/unittest2-wheel-test/lib/python2.7/site-packages/pkg_resources/__init__.py", line 1369, in safe_version
      return str(packaging.version.Version(version))
    File "/Users/marca/python/virtualenvs/unittest2-wheel-test/lib/python2.7/site-packages/pkg_resources/_vendor/packaging/version.py", line 210, in __init__
      match = self._regex.search(version)
  TypeError: expected string or buffer

  ----------------------------------------
  Failed building wheel for unittest2
  Running setup.py bdist_wheel for argparse
  Stored in directory: /Users/marca/wheelhouse
  Running setup.py bdist_wheel for six
  Stored in directory: /Users/marca/wheelhouse
  Running setup.py bdist_wheel for traceback2
  Stored in directory: /Users/marca/wheelhouse
  Running setup.py bdist_wheel for linecache2
  Stored in directory: /Users/marca/wheelhouse
Successfully built argparse six traceback2 linecache2
Failed to build unittest2
ERROR: Failed to build one or more wheels
```

https://hg.python.org/unittest2/file/d091f0086b03/setup.py#l13

---
- Bitbucket: https://bitbucket.org/pypa/setuptools/issue/409

**Comment by beamerblvd** [comment]
@msabramo:

This has bit me hard. I cannot use `unittest2` at all. I have tried to install all versions 0.8.0 through 1.1.0 to no avail. I get the same error as OP. I am completely blocked on using `unittest2` and am having to live without it.

**Comment by Avasam** [comment]
`unittest2` is Python 2 only and setuptools no longer supports Python 2. This issue is probably no longer relevant.

**Comment by abravalheri** [comment]
Thank you very much @Avasam .


I will close this issue as stale for now, but the author can re-open it with new information if necessary or create a new issue.
